resolvers/manatee-primary-resolver/src/lib.rs

```diff
@@ -204,23 +226,210 @@ impl<T: Debug> SlogValue for LogItem<T> {
+                        }
+                    }
+                    ResolverBackoffState::Unstable(current_backoff) => {
+                        debug!(
+                            thread_log,
+                            "backoff unstable; waiting for stability"
+                        );
+                        //
+                        // See large comment above for explanation of why we
+                        // wait this long
+                        //
+                        match error_rx.recv_timeout(
+                            current_backoff + *BACKOFF_RESET_THRESHOLD,
+                        ) {
+                            //
+                            // We got another error, so restart the countdown
+                            //
+                            Ok(ResolverBackoffMsg::ErrorOccurred(
+                                new_backoff,
+                            )) => {
+                                debug!(
+                                    thread_log,
+                                    "error received while waiting for \
+                                     stability; restarting wait period"
+                                );
+                                state =
+                                    ResolverBackoffState::Unstable(new_backoff);
+                                continue;
+                            }
+                            //
+                            // Timeout waiting for an error: the stability
+                            // threshold has been reached, so reset the backoff
+                            //
+                            Err(RecvTimeoutError::Timeout) => {
+                                info!(
+                                    thread_log,
+                                    "stability threshold reached; resetting backoff"
+                                );
+                                let backoff =
+                                    &mut *backoff_clone.lock().unwrap();
+                                backoff.reset();
+                                state = ResolverBackoffState::Stable
+                            }
+                            //
+                            // ResolverBackoff object was dropped, so we exit
+                            // the thread
+                            //
+                            Err(RecvTimeoutError::Disconnected) => break,
+                        }
+                    }
+                }
+            }
+            debug!(thread_log, "stability-tracking thread exiting");
+        });
+
+        ResolverBackoff {
+            backoff,
+            error_tx,
+            log,
+        }
+    }
+
+    fn next_backoff(&mut self) -> Duration {
+        let backoff = &mut *self.backoff.lock().unwrap();
+        //
+        // This should never fail because we set max_elapsed_time to `None`
+        // above.
+        //
+        let next_backoff = backoff.next_backoff().unwrap();
```

Comment:
Can we make this an `expect` just to be extra cautious?

Reply:
Done